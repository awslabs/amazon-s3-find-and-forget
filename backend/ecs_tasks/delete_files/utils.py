import time
from botocore.exceptions import ClientError


def remove_none(d: dict):
    return {k: v for k, v in d.items() if v is not None and v is not ''}


def retry_wrapper(fn, retry_wait_seconds=2, retry_factor=2, max_retries=5):
    """ Exponential back-off retry wrapper for ClientError exceptions """

    def wrapper(*args, **kwargs):
        retry_current = 0
        last_error = None

        while retry_current <= max_retries:
            try:
                return fn(*args, **kwargs)
            except ClientError as e:
                nonlocal retry_wait_seconds
                if retry_current == max_retries:
                    break
                last_error = e
                retry_current += 1
                time.sleep(retry_wait_seconds)
                retry_wait_seconds *= retry_factor

        raise last_error

    return wrapper
