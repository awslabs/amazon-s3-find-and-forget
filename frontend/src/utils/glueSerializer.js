import { isUndefined } from "./";

export const glueSerializer = tables => {
  const result = { databases: [] };

  const ARRAYSTRUCT = "array<struct>";
  const ARRAYSTRUCTPREFIX = "array<struct<";
  const ARRAYSTRUCTSUFFIX = ">>";

  const STRUCT = "struct";
  const STRUCTPREFIX = "struct<";
  const STRUCTSUFFIX = ">";

  const SCHEMA_INVALID = "Column schema is not valid";
  const ALLOWEDTYPES = [
    "bigint",
    "char",
    "double",
    "float",
    "int",
    "smallint",
    "string",
    "tinyint",
    "varchar"
  ];

  // Function to get inner children from complex type string
  // "struct<name:string,age:int>" => "name:string,age:int"
  const getInnerChildren = (str, prefix, suffix) => {
    if (!str.endsWith(suffix)) throw new Error(SCHEMA_INVALID);
    return str.substr(prefix.length).slice(0, -suffix.length);
  };

  // Function to get next nested child type from a children string
  // starting with a complex type such as struct or array
  // "struct<name:string,age:int,s:struct<n:int>>,b:string" =>
  // "struct<name:string,age:int,s:struct<n:int>>"
  const getNestedChildren = (str, nestedType) => {
    const isStruct = nestedType === STRUCT;
    const prefix = isStruct ? STRUCTPREFIX : ARRAYSTRUCTPREFIX;
    const suffix = isStruct ? STRUCTSUFFIX : ARRAYSTRUCTSUFFIX;
    let nOpenedTags = suffix.length;
    let endIndex = -1;

    const structSchema = [...str.substr(prefix.length)];
    for (let i = 0; i < structSchema.length; i++) {
      const c = structSchema[i];
      if (c === "<") nOpenedTags++;
      if (c === ">") nOpenedTags--;
      if (nOpenedTags === 0) {
        endIndex = i;
        break;
      }
    }

    if (endIndex < 0) throw new Error(SCHEMA_INVALID);
    return str.substr(0, endIndex + prefix.length + 1);
  };

  // Function to get next nested child type from a children string
  // starting with a non complex type
  // "string,a:int" => "string"
  const getNestedType = str => {
    const upperIndex = str.indexOf(",");
    return upperIndex >= 0 ? str.substr(0, upperIndex) : str;
  };

  // Function to set canBeIdentifier=false to item and its children
  // Example:
  // {
  //   name: "arr",
  //   type: "array<struct>",
  //   canBeIdentifier: false,
  //   children: [
  //     { name: "field", type: "int", canBeIdentifier: true },
  //     { name: "n", type: "string", canBeIdentifier: true }
  //   ]
  // }
  // =>
  // {
  //   name: "arr",
  //   type: "array<struct>",
  //   canBeIdentifier: false,
  //   children: [
  //     { name: "field", type: "int", canBeIdentifier: false },
  //     { name: "n", type: "string", canBeIdentifier: false }
  //   ]
  // }
  const setNoIdentifierToNodeAndItsChildren = node => {
    node.canBeIdentifier = false;
    if (node.children)
      node.children.forEach(c => setNoIdentifierToNodeAndItsChildren(c));
  };

  // Function to map Columns from AWS Glue schema to tree
  // Example 1:
  // { Name: "Name", Type: "int" } =>
  // { name: "Name", type: "int", canBeIdentifier: true }
  // Example 2:
  // { Name: "complex", Type: "struct<a:string,b:struct<c:int>>"} =>
  // { name: "complex", type: "struct", children: [
  //    { name: "a", type: "string", canBeIdentifier: false},
  //    { name: "b", type: "struct", children: [
  //      { name: "c", type: "int", canBeIdentifier: false}
  //    ], canBeIdentifier: false}
  // ], canBeIdentifier: false}
  const columnMapper = c => {
    let prefix, suffix;
    let resultType = c.Type;
    let hasChildren = false;

    if (c.Type.startsWith(ARRAYSTRUCTPREFIX)) {
      resultType = ARRAYSTRUCT;
      prefix = ARRAYSTRUCTPREFIX;
      suffix = ARRAYSTRUCTSUFFIX;
      hasChildren = true;
    } else if (c.Type.startsWith(STRUCTPREFIX)) {
      resultType = STRUCT;
      prefix = STRUCTPREFIX;
      suffix = STRUCTSUFFIX;
      hasChildren = true;
    }

    const result = {
      name: c.Name,
      type: resultType,
      canBeIdentifier: !isUndefined(c.canBeIdentifier)
        ? c.canBeIdentifier
        : ALLOWEDTYPES.includes(resultType)
    };

    if (hasChildren) {
      result.children = [];
      let childrenToParse = getInnerChildren(c.Type, prefix, suffix);

      while (childrenToParse.length > 0) {
        const sep = ":";
        const name = childrenToParse.substr(0, childrenToParse.indexOf(sep));
        const rest = childrenToParse.substr(name.length + sep.length);
        const nestedType = rest.startsWith(STRUCTPREFIX)
          ? STRUCT
          : rest.startsWith(ARRAYSTRUCTPREFIX)
          ? ARRAYSTRUCT
          : "other";

        const type =
          nestedType === "other"
            ? getNestedType(rest)
            : getNestedChildren(rest, nestedType);

        result.children.push(
          columnMapper({
            Name: name,
            Type: type,
            canBeIdentifier: ALLOWEDTYPES.includes(type)
          })
        );
        childrenToParse = childrenToParse.substr(
          name.length + sep.length + type.length
        );
        if (childrenToParse.startsWith(","))
          childrenToParse = childrenToParse.substr(1);
      }
      if (resultType !== STRUCT) setNoIdentifierToNodeAndItsChildren(result);
    }

    return result;
  };

  tables.forEach(table => {
    const parquetTables = table.TableList.filter(
      x =>
        x.StorageDescriptor.SerdeInfo.SerializationLibrary ===
        "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
    );

    if (parquetTables.length > 0)
      result.databases.push({
        name: parquetTables[0].DatabaseName,
        tables: parquetTables.map(t => ({
          name: t.Name,
          columns: t.StorageDescriptor.Columns.map(columnMapper)
        }))
      });
  });
  return result;
};

export const bucketMapper = tables => {
  const result = {};

  tables.forEach(
    t =>
      (result[`${t.Table.DatabaseName}/${t.Table.Name}`] = {
        bucket: t.Table.StorageDescriptor.Location.split("/")[2],
        location: t.Table.StorageDescriptor.Location
      })
  );

  return result;
};
