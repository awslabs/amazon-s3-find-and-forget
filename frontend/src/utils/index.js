export const formatErrorMessage = e => {
  if (e && e.response) {
    if (e.response.data && e.response.data.Message) {
      return e.response.data.Message;
    } else {
      return `Unknown error (${e.response.status} status code)`;
    }
  }

  return e ? e.toString() : "Unknown error";
};

export const isEmpty = x =>
  x === null ||
  isUndefined(x) ||
  (Array.isArray(x)
    ? x.length === 0
    : typeof x === "string"
    ? x.trim() === ""
    : false);

export const isIdValid = x => {
  const idRegex = /^[a-zA-Z0-9]+$/;
  return idRegex.test(x);
};

export const isRoleArnValid = x => {
  const arnRegex = /^arn:(aws[a-zA-Z-]*)?:iam::\d{12}:role\/S3F2DataAccessRole$/;
  return arnRegex.test(x);
};

export const sortBy = (obj, key) =>
  obj.sort((a, b) => (a[key] > b[key] ? 1 : a[key] < b[key] ? -1 : 0));

export const daysSinceDateTime = x => {
  const now = new Date();
  const from = x ? new Date(x * 1000) : now;
  const aDay = 24 * 60 * 60 * 1000;
  return parseInt((now - from) / aDay, 10);
};

export const formatDateTime = x => (x ? new Date(x * 1000).toUTCString() : "-");

export const formatFileSize = x => {
  const units = ["bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];

  if (x === 1) return "1 byte";

  let l = 0;
  let n = parseInt(x, 10) || 0;

  while (n >= 1024 && ++l) n = n / 1024;

  return n.toFixed(n < 10 && l > 0 ? 1 : 0) + " " + units[l];
};

export const isUndefined = x => typeof x === "undefined";

export const withDefault = (x, formatter = () => x) =>
  isEmpty(x) ? "-" : formatter(x);

export const successJobClass = status => {
  switch (status) {
    case "COMPLETED":
      return "success";
    case "RUNNING":
    case "QUEUED":
    case "FORGET_COMPLETED_CLEANUP_IN_PROGRESS":
      return "info";
    default:
      return "error";
  }
};

export const last = x => (x.length > 0 ? x[x.length - 1] : null);

const trimTrailingSlash = x => x.replace(/\/+$/, "");
const trimLeadingSlash = x => x.replace(/\/+$/, "");

export const repoUrl = x => {
  const baseUrl = trimTrailingSlash(
    process.env.REACT_APP_REPO_URL.replace("git+", "").replace(".git", "")
  );
  const path = trimLeadingSlash(x);
  return trimTrailingSlash(`${baseUrl}/${path}`);
};

export const docsUrl = x => repoUrl(`blob/master/docs/${trimLeadingSlash(x)}`);

export const findMin = (arr, key) =>
  arr.reduce((prev, curr) => (prev[key] < curr[key] ? prev : curr));
