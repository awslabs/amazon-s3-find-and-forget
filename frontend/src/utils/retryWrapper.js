const MAX_RETRIES = 3;
const RETRY_START = 1000;

export const retryWrapper = (p, timeout, retryN) =>
  new Promise((resolve, reject) =>
    p()
      .then(resolve)
      .catch(e => {
        const retryable = e?.response?.status !== 400;
        if (retryN === MAX_RETRIES || !retryable) return reject(e);
        const t = (timeout || RETRY_START / 2) * 2;
        const r = (retryN || 0) + 1;
        console.log(`Retry n. ${r} in ${t / 1000}s...`);
        setTimeout(() => retryWrapper(p, t, r).then(resolve).catch(reject), t);
      })
  );
