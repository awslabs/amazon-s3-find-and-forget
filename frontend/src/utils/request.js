import Amplify, { API, Auth } from "aws-amplify";
import { retryWrapper } from "./retryWrapper";

const settings = window.s3f2Settings || {};
const region = settings.region || "eu-west-1";

const getRegionalConfig = serviceName => ({
  name: serviceName,
  endpoint: `https://${serviceName}.${region}.amazonaws.com`,
  region,
  service: serviceName
});

Amplify.configure({
  Auth: {
    identityPoolId: settings.cognitoIdentityPool,
    region,
    mandatorySignIn: true,
    userPoolId: settings.cognitoUserPoolId,
    userPoolWebClientId: settings.cognitoUserPoolClientId
  },
  API: {
    endpoints: [
      {
        name: "apiGateway",
        endpoint: `${settings.apiUrl}v1/`,
        region,
        custom_header: async () => {
          const session = await Auth.currentSession();
          const token = session.getIdToken().getJwtToken();
          return { Authorization: `Bearer ${token}` };
        }
      },
      getRegionalConfig("glue"),
      getRegionalConfig("sts")
    ]
  }
});

const apiWrapper = (api, endpoint, options) => {
  const { data, headers, method, response } = options;
  const reqOptions = {
    headers: headers || { "Content-Type": "application/json" },
    response: response || false
  };
  if (data) reqOptions.body = data;
  return retryWrapper(() => API[method || "get"](api, endpoint, reqOptions));
};

export const apiGateway = (endpoint, options = {}) =>
  apiWrapper("apiGateway", endpoint, options);

export const glueGateway = (endpointName, data) =>
  apiWrapper("glue", "", {
    data,
    headers: {
      "Content-Type": "application/x-amz-json-1.1",
      "X-Amz-Target": `AWSGlue.${endpointName}`
    },
    method: "post"
  });

export const stsGateway = endpoint =>
  apiWrapper("sts", endpoint, { method: "post" });
