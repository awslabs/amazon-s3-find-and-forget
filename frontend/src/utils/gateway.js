import { apiGateway, glueGateway, stsGateway } from "./request";

const getPaginatedList = async (endpoint, key, pageSize) => {
  const all = [];
  let watermark = undefined;
  let contentLength = 0;

  while (true) {
    let qs = `?page_size=${pageSize || 10}`;
    if (watermark) qs += `&start_at=${watermark}`;
    const page = await apiGateway(`${endpoint}${qs}`, { response: true });
    all.push(...page.data[key]);
    contentLength += parseInt(page.headers["content-length"], 10);

    if (page.data.NextStart)
      watermark = encodeURIComponent(page.data.NextStart);
    else break;
  }

  return { contentLength, response: all };
};

export default {
  deleteDataMapper(id) {
    return apiGateway(`data_mappers/${id}`, { method: "del" });
  },

  deleteQueueMatches(matches) {
    return apiGateway(`queue/matches`, {
      method: "del",
      data: {
        Matches: matches
      }
    });
  },

  enqueue(id, dataMappers) {
    return apiGateway(`queue`, {
      method: "patch",
      data: { MatchId: id, DataMappers: dataMappers }
    });
  },

  getAccountId() {
    return stsGateway("/?Action=GetCallerIdentity&Version=2011-06-15");
  },

  async getDataMappers() {
    const dataMappers = await getPaginatedList(
      "data_mappers",
      "DataMappers",
      100
    );
    return { DataMappers: dataMappers.response };
  },

  async getGlueDatabases() {
    const all = [];
    let watermark = undefined;
    while (true) {
      const body = watermark ? { NextToken: watermark } : {};
      const page = await glueGateway("GetDatabases", body);
      all.push(...page.DatabaseList);

      if (page.NextToken) watermark = page.NextToken;
      else break;
    }
    return { DatabaseList: all };
  },

  getGlueTable(database, table) {
    return glueGateway("GetTable", {
      DatabaseName: database,
      Name: table
    });
  },

  async getGlueTables(database) {
    const all = [];
    let watermark = undefined;
    while (true) {
      const body = Object.assign(watermark ? { NextToken: watermark } : {}, {
        DatabaseName: database
      });
      const page = await glueGateway("GetTables", body);
      all.push(...page.TableList);

      if (page.NextToken) watermark = page.NextToken;
      else break;
    }
    return { TableList: all };
  },

  getJob(jobId) {
    return apiGateway(`jobs/${jobId}`);
  },

  getLastJob() {
    return apiGateway(`jobs?page_size=1`);
  },

  async getJobs() {
    const jobs = await getPaginatedList("jobs", "Jobs");
    return { Jobs: jobs.response };
  },

  async getJobEvents(
    jobId,
    watermark = undefined,
    page_size = 20,
    filters = []
  ) {
    let qs = `?page_size=${page_size}`;
    if (watermark) {
      qs += `&start_at=${encodeURIComponent(watermark)}`;
    }
    if (filters.length > 0) {
      qs +=
        "&" +
        filters
          .map(
            f => `filter=${encodeURIComponent(f.key + f.operator + f.value)}`
          )
          .join("&");
    }
    return await apiGateway(`jobs/${jobId}/events${qs}`);
  },

  async getAllJobEvents(jobId) {
    let watermark = "0";
    let data = [];
    while (watermark) {
      let jobEventsList = await this.getJobEvents(jobId, watermark, 1000);
      data = data.concat(jobEventsList.JobEvents);
      watermark = jobEventsList.NextStart;
    }

    return data;
  },

  async getQueue() {
    const { response, contentLength } = await getPaginatedList(
      "queue",
      "MatchIds",
      500
    );
    return { MatchIds: response, ContentLength: contentLength };
  },

  getSettings() {
    return apiGateway("settings");
  },

  processQueue() {
    return apiGateway("queue", { method: "del" });
  },

  putDataMapper(id, db, table, columns, roleArn, deleteOldVersions, format) {
    return apiGateway(`data_mappers/${id}`, {
      method: "put",
      data: {
        DataMapperId: id,
        Columns: columns,
        QueryExecutor: "athena",
        QueryExecutorParameters: {
          DataCatalogProvider: "glue",
          Database: db,
          Table: table
        },
        Format: format,
        RoleArn: roleArn,
        DeleteOldVersions: deleteOldVersions
      }
    });
  }
};
