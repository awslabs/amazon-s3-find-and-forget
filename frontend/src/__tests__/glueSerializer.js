import { bucketMapper, glueSerializer } from "../utils/glueSerializer";

const tableMaker = ({
  dbname,
  tablename,
  partitions = [],
  columns = [],
  location,
  serde = {
    Parameters: { "serialization.format": "1" },
    SerializationLibrary:
      "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe"
  }
}) => ({
  CreateTime: 1.571744695e9,
  DatabaseName: dbname,
  IsRegisteredWithLakeFormation: false,
  LastAccessTime: 0.0,
  Name: tablename,
  Owner: "hadoop",
  Parameters: {
    EXTERNAL: "TRUE",
    has_encrypted_data: "false",
    transient_lastDdlTime: "1571744695"
  },
  PartitionKeys: partitions,
  Retention: 0,
  StorageDescriptor: {
    BucketColumns: [],
    Columns: columns,
    Compressed: false,
    InputFormat: "org.apache.hadoop.mapred.TextInputFormat",
    Location: location,
    NumberOfBuckets: -1,
    OutputFormat: "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
    Parameters: {},
    SerdeInfo: serde,
    SkewedInfo: {
      SkewedColumnNames: [],
      SkewedColumnValueLocationMaps: {},
      SkewedColumnValues: []
    },
    SortColumns: [],
    StoredAsSubDirectories: false
  },
  TableType: "EXTERNAL_TABLE",
  UpdateTime: 1.572537815e9
});

const table1 = tableMaker({
  dbname: "db2",
  tablename: "table1",
  partitions: [{ Name: "product_category", Type: "string" }],
  columns: [{ Name: "customer_id", Type: "string" }],
  location: "s3://my-s3-bucket/parquet/"
});

const table2 = tableMaker({
  dbname: "db2",
  tablename: "table2",
  columns: [{ Name: "author", Type: "string" }],
  location: "s3://my-s3-bucket/parquet2/"
});

const table3 = tableMaker({
  dbname: "db3",
  tablename: "table5",
  columns: [{ Name: "customer_id", Type: "string" }],
  partitions: [{ Name: "product_category", Type: "string" }],
  location: "s3://my-s3-bucket/parquet56/"
});

const jsonTable = tableMaker({
  dbname: "db3",
  tablename: "table3",
  columns: [
    { Name: "id", Type: "string" },
    { Name: "type", Type: "string" },
    { Name: "repo", Type: "struct<id:int,name:string,url:string>" },
    { Name: "public", Type: "boolean" },
    { Name: "created_at", Type: "string" }
  ],
  partitions: [
    { Name: "partition_0", Type: "string" },
    { Name: "partition_1", Type: "string" },
    { Name: "partition_2", Type: "string" }
  ],
  location:
    "s3://aws-glue-datasets-eu-west-1/examples/githubarchive/month/data/",
  serde: {
    Parameters: {
      paths: "created_at,id,public,repo,type"
    },
    SerializationLibrary: "org.openx.data.jsonserde.JsonSerDe"
  }
});

const complexColumnsTable = tableMaker({
  dbname: "db4",
  tablename: "complex",
  columns: [
    { Name: "id", Type: "string" },
    { Name: "type", Type: "string" },
    { Name: "repo", Type: "struct<id:int,name:string,url:string>" },
    { Name: "simplearr", Type: "array<int>" },
    { Name: "arr", Type: "array<struct<field:int,n:string>>" },
    {
      Name: "nestedstruct",
      Type:
        "struct<a:int,b:string,c:struct<d:int,e:struct<f:int>>,g:struct<h:string>>"
    },
    {
      Name: "structandarr",
      Type:
        "struct<a:int,b:string,c:struct<d:int,e:struct<f:int>>,g:struct<h:string>,i:array<struct<l:int,m:struct<n:string>>>>"
    }
  ],
  partitions: [],
  location: "s3://my-s3-bucket/parquet/"
});

test("it should serialize dbs and tables", () => {
  expect(
    glueSerializer([
      { TableList: [] },
      { TableList: [table1, table2] },
      { TableList: [table3, jsonTable] },
      { TableList: [complexColumnsTable] }
    ])
  ).toMatchSnapshot();
});

test("it should throw error when deserializing a invalid table", () => {
  const brokenTable = tableMaker({
    dbname: "db5",
    tablename: "broken",
    columns: [
      {
        Name: "brokenstruct",
        Type: "struct<a:int"
      }
    ],
    partitions: [],
    location: "s3://my-s3-bucket/parquet/"
  });

  const failureScenario = [{ TableList: [brokenTable] }];
  try {
    glueSerializer(failureScenario);
  } catch (e) {
    expect(e).toBeInstanceOf(Error);
    expect(e).toHaveProperty("message", "Column schema is not valid");
  }
});

test("it should group buckets by table", () => {
  const getTableResponseArray = [table1, table2, table3, jsonTable].map(t => ({
    Table: t
  }));

  expect(bucketMapper(getTableResponseArray)).toEqual({
    "db2/table1": {
      bucket: "my-s3-bucket",
      location: "s3://my-s3-bucket/parquet/"
    },
    "db2/table2": {
      bucket: "my-s3-bucket",
      location: "s3://my-s3-bucket/parquet2/"
    },
    "db3/table5": {
      bucket: "my-s3-bucket",
      location: "s3://my-s3-bucket/parquet56/"
    },
    "db3/table3": {
      bucket: "aws-glue-datasets-eu-west-1",
      location:
        "s3://aws-glue-datasets-eu-west-1/examples/githubarchive/month/data/"
    }
  });
});
