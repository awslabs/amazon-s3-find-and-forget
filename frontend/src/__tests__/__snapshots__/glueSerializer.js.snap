// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should serialize dbs and tables 1`] = `
Object {
  "databases": Array [
    Object {
      "name": "db2",
      "tables": Array [
        Object {
          "columns": Array [
            Object {
              "canBeIdentifier": true,
              "name": "customer_id",
              "type": "string",
            },
          ],
          "format": "parquet",
          "name": "table1",
        },
        Object {
          "columns": Array [
            Object {
              "canBeIdentifier": true,
              "name": "author",
              "type": "string",
            },
          ],
          "format": "parquet",
          "name": "table2",
        },
      ],
    },
    Object {
      "name": "db3",
      "tables": Array [
        Object {
          "columns": Array [
            Object {
              "canBeIdentifier": true,
              "name": "customer_id",
              "type": "string",
            },
          ],
          "format": "parquet",
          "name": "table5",
        },
        Object {
          "columns": Array [
            Object {
              "canBeIdentifier": true,
              "name": "id",
              "type": "string",
            },
            Object {
              "canBeIdentifier": true,
              "name": "type",
              "type": "string",
            },
            Object {
              "canBeIdentifier": false,
              "children": Array [
                Object {
                  "canBeIdentifier": true,
                  "name": "id",
                  "type": "int",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "name",
                  "type": "string",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "url",
                  "type": "string",
                },
              ],
              "name": "repo",
              "type": "struct",
            },
            Object {
              "canBeIdentifier": false,
              "name": "public",
              "type": "boolean",
            },
            Object {
              "canBeIdentifier": true,
              "name": "created_at",
              "type": "string",
            },
          ],
          "format": "json",
          "name": "table3",
        },
      ],
    },
    Object {
      "name": "db4",
      "tables": Array [
        Object {
          "columns": Array [
            Object {
              "canBeIdentifier": true,
              "name": "id",
              "type": "string",
            },
            Object {
              "canBeIdentifier": true,
              "name": "type",
              "type": "string",
            },
            Object {
              "canBeIdentifier": false,
              "children": Array [
                Object {
                  "canBeIdentifier": true,
                  "name": "id",
                  "type": "int",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "name",
                  "type": "string",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "url",
                  "type": "string",
                },
              ],
              "name": "repo",
              "type": "struct",
            },
            Object {
              "canBeIdentifier": false,
              "name": "simplearr",
              "type": "array<int>",
            },
            Object {
              "canBeIdentifier": false,
              "children": Array [
                Object {
                  "canBeIdentifier": false,
                  "name": "field",
                  "type": "int",
                },
                Object {
                  "canBeIdentifier": false,
                  "name": "n",
                  "type": "string",
                },
              ],
              "name": "arr",
              "type": "array<struct>",
            },
            Object {
              "canBeIdentifier": false,
              "children": Array [
                Object {
                  "canBeIdentifier": true,
                  "name": "a",
                  "type": "int",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "b",
                  "type": "string",
                },
                Object {
                  "canBeIdentifier": false,
                  "children": Array [
                    Object {
                      "canBeIdentifier": true,
                      "name": "d",
                      "type": "int",
                    },
                    Object {
                      "canBeIdentifier": false,
                      "children": Array [
                        Object {
                          "canBeIdentifier": true,
                          "name": "f",
                          "type": "int",
                        },
                      ],
                      "name": "e",
                      "type": "struct",
                    },
                  ],
                  "name": "c",
                  "type": "struct",
                },
                Object {
                  "canBeIdentifier": false,
                  "children": Array [
                    Object {
                      "canBeIdentifier": true,
                      "name": "h",
                      "type": "string",
                    },
                  ],
                  "name": "g",
                  "type": "struct",
                },
              ],
              "name": "nestedstruct",
              "type": "struct",
            },
            Object {
              "canBeIdentifier": false,
              "children": Array [
                Object {
                  "canBeIdentifier": true,
                  "name": "a",
                  "type": "int",
                },
                Object {
                  "canBeIdentifier": true,
                  "name": "b",
                  "type": "string",
                },
                Object {
                  "canBeIdentifier": false,
                  "children": Array [
                    Object {
                      "canBeIdentifier": true,
                      "name": "d",
                      "type": "int",
                    },
                    Object {
                      "canBeIdentifier": false,
                      "children": Array [
                        Object {
                          "canBeIdentifier": true,
                          "name": "f",
                          "type": "int",
                        },
                      ],
                      "name": "e",
                      "type": "struct",
                    },
                  ],
                  "name": "c",
                  "type": "struct",
                },
                Object {
                  "canBeIdentifier": false,
                  "children": Array [
                    Object {
                      "canBeIdentifier": true,
                      "name": "h",
                      "type": "string",
                    },
                  ],
                  "name": "g",
                  "type": "struct",
                },
                Object {
                  "canBeIdentifier": false,
                  "children": Array [
                    Object {
                      "canBeIdentifier": false,
                      "name": "l",
                      "type": "int",
                    },
                    Object {
                      "canBeIdentifier": false,
                      "children": Array [
                        Object {
                          "canBeIdentifier": false,
                          "name": "n",
                          "type": "string",
                        },
                      ],
                      "name": "m",
                      "type": "struct",
                    },
                  ],
                  "name": "i",
                  "type": "array<struct>",
                },
              ],
              "name": "structandarr",
              "type": "struct",
            },
          ],
          "format": "parquet",
          "name": "complex",
        },
      ],
    },
  ],
}
`;
